libafl/src/observers/cmp.rs

```diff
@@ -725,14 +944,74 @@ impl Debug for AFLppCmpVals {
     }
 }
 
-/// The AFL++ `cmp_map` struct, use with `ForkserverCmpObserver`
+impl AFLppCmpVals {
+    #[must_use]
+    /// Reference comparison values as comparison operands
+    pub fn operands(&self) -> &[[AFLppCmpOperands; AFL_CMP_MAP_H]; AFL_CMP_MAP_W] {
+        unsafe { &self.operands }
+    }
+
+    #[must_use]
+    /// Mutably reference comparison values as comparison operands
+    pub fn operands_mut(&mut self) -> &[[AFLppCmpOperands; AFL_CMP_MAP_H]; AFL_CMP_MAP_W] {
+        unsafe { &mut self.operands }
+    }
+
+    #[must_use]
+    /// Reference comparison values as comparison function operands
+    pub fn fn_operands(&self) -> &[[AFLppCmpFnOperands; AFL_CMP_MAP_RTN_H]; AFL_CMP_MAP_W] {
+        unsafe { &self.fn_operands }
+    }
+
+    #[must_use]
+    /// Mutably reference comparison values as comparison function operands
+    pub fn fn_operands_mut(&mut self) -> &[[AFLppCmpFnOperands; AFL_CMP_MAP_RTN_H]; AFL_CMP_MAP_W] {
+        unsafe { &mut self.fn_operands }
+    }
+}
+
 #[derive(Debug, Clone, Copy)]
 #[repr(C, packed)]
+/// Comparison map compatible with AFL++ cmplog instrumentation
 pub struct AFLppCmpMap {
     headers: [AFLppCmpHeader; AFL_CMP_MAP_W],
     vals: AFLppCmpVals,
 }
 
+impl AFLppCmpMap {
+    #[must_use]
+    /// Instantiate a new boxed zeroed `AFLppCmpMap`. This should be used to create a new
+    /// map, because it is so large it cannot be allocated on the stack with default
+    /// runtime configuration.
+    pub fn boxed() -> Box<Self> {
+        unsafe { Box::from_raw(alloc_zeroed(Layout::new::<AFLppCmpMap>()) as *mut AFLppCmpMap) }
```

Comment:
Stupid & unrelated question: could this be a normal `Box::new(AFLppCmpMap)` of sorts to get around unsafe? Maybe `AFLppCmpFnOperands` would need a default for that?

Reply:
I tried a bunch of stuff and was getting stack overflows for a lot of things but I don't recall if this is one of the things I tried -- I'll give it a shot.

Reply:
Ah yeah, can't be done :( If you do `Box::new(AFLppCmpMap::new())` for example, it constructs it on the stack (causing a stack overflow) and then would memcpy it onto the heap -- using `alloc` or `alloc_zeroed` afaik is the only way to directly allocate it on the stack.

Reply:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021 even doing it this way causes a stack overflow, actually :/

Reply:
Your link doesn't contain code 😄
But all right sounds good

Reply:
https://users.rust-lang.org/t/how-to-boxed-struct-with-large-size-without-stack-overflow/94961/15

Yes looks like that's what you need.
Can you add a comment explaining this? 🙃